toucan_connectors/snowflake/snowflake_connector.py

```diff
@@ -282,174 +306,134 @@ def _refresh_oauth_token(self):
+            assert query_result is not None
+            # snowflake typing is incomplete for DictCursor
+            results: list[dict] = query_result.fetchall()  # type:ignore[assignment]
+            return pd.DataFrame(results) if as_df else results
+
+    def _describe_query(self, query: str) -> dict[str, str]:
+        with self._get_connection() as conn:
+            curs = conn.cursor(SfDictCursor)
+            return {r.name: _TYPE_CODE_MAPPING[r.type_code] for r in curs.describe(query)}
+
+    def _get_warehouses(self, warehouse_name: str | None = None) -> list[str]:
+        query = 'SHOW WAREHOUSES'
+        if warehouse_name:
+            query += f"LIKE '{warehouse_name}'"
+
+        result = self._execute_query(query, warehouse=warehouse_name)
+        return result['name'].to_list() if 'name' in result.columns else []
+
+    def _get_databases(self, database_name: str | None = None) -> list[str]:
+        query = 'SHOW DATABASES'
+        if database_name:
+            query += f"LIKE '{database_name}'"
+
+        result = self._execute_query(query, database=database_name)
+        return result['name'].to_list() if 'name' in result.columns else []
 
     def _fetch_data(
         self,
         data_source: SnowflakeDataSource,
-        offset: Optional[int] = None,
-        limit: Optional[int] = None,
-        get_row_count: bool = False,
+        offset: int | None = None,
+        limit: int | None = None,
     ) -> pd.DataFrame:
         data_source = self._set_warehouse(data_source)
-        with self._get_connection(data_source.database, data_source.warehouse) as connection:
-            result = SnowflakeCommon().fetch_data(
-                connection, data_source, offset, limit, get_row_count
-            )
-        return result
+
+        prepared_query, prepared_params = SqlQueryHelper.prepare_limit_query(
+            data_source.query, data_source.parameters, offset=offset, limit=limit
+        )
+        return self._execute_query(prepared_query, prepared_params)
+
+    def _retrieve_data(self, data_source: SnowflakeDataSource) -> pd.DataFrame:
+        return self._fetch_data(data_source)
 
     def get_slice(
         self,
         data_source: SnowflakeDataSource,
-        permissions: Optional[dict] = None,
+        permissions: dict | None = None,
         offset: int = 0,
-        limit: Optional[int] = None,
-        get_row_count: Optional[bool] = False,
+        limit: int | None = None,
+        get_row_count: bool | None = False,
     ) -> DataSlice:
-        data_source = self._set_warehouse(data_source)
-        with self._get_connection(data_source.database, data_source.warehouse) as connection:
-            result = SnowflakeCommon().get_slice(
-                connection, data_source, offset=offset, limit=limit, get_row_count=get_row_count
-            )
-        return result
+        # FIXME: are permissions ignored for the snowflake connector ?
```

Comment:
Apparently yes − but not with the new way of applying permissions (as FilterSteps) so maybe we can ignore it ? or copy the code from `toucan_connector.py` otherwise

Reply:
That's true! I'd rather leave it as is, since they were also ignored before